dali/kernels/algebra/linear_transformation_gpu.h

```diff
@@ -0,0 +1,198 @@
+    for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+      // Iterate over channels: `o` over output, `i` over input
+      // It's a plain matrix multiplication
+      for (int o = 0; o < N; o++) {
+        OutputType val = 0;
+        for (int i = 0; i < M; i++) {
+          val += sample.transformation_matrix.at(o, i) * in(x, y, i);
+        }
+        out(x, y, o) = val;
+      }
+    }
+  }
+}
+
+
+template <typename OutputType, typename InputType, size_t N, size_t M, size_t spatial_ndims>
+class LinearTransformationGpu {
+ private:
+  static constexpr auto ndims_ = spatial_ndims + 1;
+  using mat = ::dali::mat<N, M, detail::TransformationMatrixElementType>;
+  using BlockDesc = kernels::BlockDesc<spatial_ndims>;
+  using SampleDescriptor = detail::SampleDescriptor<OutputType, InputType, N, M, spatial_ndims>;
+
+  std::vector<SampleDescriptor> sample_descriptors_;
+
+ public:
+  BlockSetup<spatial_ndims,
+          spatial_ndims /* Assumed, that the channel dim is the last dim */> block_setup_;
+
+
+  KernelRequirements
+  Setup(KernelContext &context, const InListGPU<InputType, ndims_> &in,
+        const std::vector<mat> &transformation_matrices,
+        const std::vector<Roi<spatial_ndims>> &rois = {}) {
+    DALI_ENFORCE(rois.empty() || rois.size() == static_cast<size_t>(in.num_samples()),
+                 "Provide ROIs either for all or none input tensors");
+    DALI_ENFORCE([=]() -> bool {
+        for (const auto &roi : rois) {
+          if (!all_coords(roi.hi >= roi.lo))
+            return false;
+        }
+        return true;
+    }(), "One or more regions of interests are invalid");
+    DALI_ENFORCE([=]() -> bool {
+        auto ref_shape = in.shape[0][ndims_ - 1];
+        for (int i = 0; i < in.num_samples(); i++) {
+          if (in.shape[i][ndims_ - 1] != ref_shape) {
+            return false;
+          }
+        }
+        return true;
+    }(), "Number of channels for all images in batch must be equal");
+
+    auto adjusted_rois = AdjustRoi(rois, in.shape);
+    auto nchannels_out = N;
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    TensorListShape<ndims_> output_shape(ShapeFromRoi(adjusted_rois, nchannels_out));
+    block_setup_.SetupBlocks(output_shape, true);
+    se.add<SampleDescriptor>(AllocType::GPU, in.num_samples());
+    se.add<BlockDesc>(AllocType::GPU, block_setup_.Blocks().size());
+    req.output_shapes = {output_shape};
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+
+  void Run(KernelContext &context, const OutListGPU<OutputType, spatial_ndims + 1> &out,
+           const InListGPU<InputType, spatial_ndims + 1> &in,
+           const std::vector<::dali::mat<N, M, detail::TransformationMatrixElementType>> &tmatrices,
+           const std::vector<Roi<spatial_ndims>> &rois = {}) {
+    auto sample_descs = detail::CreateSampleDescriptors<OutputType, InputType, N, M, spatial_ndims>
```

Comment:
Sample descriptors must be aware of ROIs.

Reply:
Done